File: problems/3113_find_the_number_of_subarrays_where_boundary_elements_are_maximum/solution.py
from typing import List


class Solution:
    def numberOfSubarrays(self, nums: List[int]) -> int:
        res = 0
        stack = []
        for x in nums:
            while stack and stack[-1][0] < x:
                stack.pop()
            if stack and stack[-1][0] == x:
                stack[-1] = (x, stack[-1][1] + 1)
                res += stack[-1][1]
            else:
                stack.append((x, 1))
                res += 1
        return res

File: problems/3113_find_the_number_of_subarrays_where_boundary_elements_are_maximum/Solution.java
import java.util.*;

class Solution {
    public long numberOfSubarrays(int[] nums) {
        long res = 0;
        Deque<long[]> stack = new ArrayDeque<>();
        for (int x : nums) {
            while (!stack.isEmpty() && stack.peek()[0] < x) stack.pop();
            if (!stack.isEmpty() && stack.peek()[0] == x) {
                stack.peek()[1]++;
                res += stack.peek()[1];
            } else {
                stack.push(new long[]{x, 1});
                res++;
            }
        }
        return res;
    }
}
